tests/unit/test_transfer_manager.py

```diff
@@ -0,0 +1,315 @@
+        FAKE_RESULT = "nothing to see here"
+
+        for blob_mock, _ in BLOB_FILE_PAIRS:
+            blob_mock.download_to_file.return_value = FAKE_RESULT
+
+        results = transfer_manager.download_many(
+            BLOB_FILE_PAIRS, download_kwargs=DOWNLOAD_KWARGS
+        )
+        for (mock_blob, file) in BLOB_FILE_PAIRS:
+            mock_blob.download_to_file.assert_any_call(file, **DOWNLOAD_KWARGS)
+        for result in results:
+            self.assertEqual(result, FAKE_RESULT)
+
+    def test_download_many_passes_concurrency_options(self):
+        BLOB_FILE_PAIRS = [
+            (mock.Mock(), tempfile.TemporaryFile()),
+            (mock.Mock(), tempfile.TemporaryFile()),
+        ]
+        MAX_WORKERS = 7
+        DEADLINE = 10
+        with mock.patch(
+            "concurrent.futures.ThreadPoolExecutor"
+        ) as pool_patch, mock.patch("concurrent.futures.wait") as wait_patch:
+            transfer_manager.download_many(
+                BLOB_FILE_PAIRS, max_workers=MAX_WORKERS, deadline=DEADLINE
+            )
+            pool_patch.assert_called_with(max_workers=MAX_WORKERS)
+            wait_patch.assert_called_with(
+                mock.ANY, timeout=DEADLINE, return_when=mock.ANY
+            )
+
+    def test_download_many_suppresses_exceptions(self):
+        BLOB_FILE_PAIRS = [(mock.Mock(), "file_a.txt"), (mock.Mock(), "file_b.txt")]
+        for mock_blob, _ in BLOB_FILE_PAIRS:
+            mock_blob.download_to_filename.side_effect = ConnectionError()
+
+        results = transfer_manager.download_many(BLOB_FILE_PAIRS)
+        for result in results:
+            self.assertEqual(type(result), ConnectionError)
+
+    def test_download_many_raises_exceptions(self):
+        BLOB_FILE_PAIRS = [(mock.Mock(), "file_a.txt"), (mock.Mock(), "file_b.txt")]
+        for mock_blob, _ in BLOB_FILE_PAIRS:
+            mock_blob.download_to_filename.side_effect = ConnectionError()
+
+        transfer_manager.download_many(BLOB_FILE_PAIRS)
+        with self.assertRaises(ConnectionError):
+            transfer_manager.download_many(BLOB_FILE_PAIRS, raise_exception=True)
+
+    def test_download_chunks_concurrently_to_file(self):
+        BLOB_CONTENTS = b"1234567812345678A"
+        blob = mock.Mock()
+        blob.size = len(BLOB_CONTENTS)
+        blob.generation = None
+
+        FAKE_ENCODING = "fake-gzip"
+        DOWNLOAD_KWARGS = {"accept-encoding": FAKE_ENCODING}
+
+        def fake_download_to_file(file_obj, start, end, **kwargs):
+            file_obj.write(BLOB_CONTENTS[start : end + 1])
+            self.assertEqual(kwargs, DOWNLOAD_KWARGS)
+
+        blob.download_to_file = fake_download_to_file
+
+        file_obj = io.BytesIO()
+
+        transfer_manager.download_chunks_concurrently_to_file(
+            blob, file_obj, chunk_size=4, download_kwargs=DOWNLOAD_KWARGS
+        )
+
+        # Generation wasn't set, so reload should have been called.
+        blob.reload.assert_called_with()
```

Comment:
Nice test here!! Maybe we want to mention in the docstrings that `blob.reload()` is called if the generation or size is not set on the blob

Reply:
Good call.